AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -211,49 +212,72 @@ + (Class)layerClass
 
 + (void)scheduleNodeForRecursiveDisplay:(ASDisplayNode *)node
 {
+
   static ASDN::RecursiveMutex __displaySchedulerLock;
-  static NSMutableArray *__nodesToDisplay = nil;
-  static BOOL __displayScheduled = NO;
-  
-  BOOL scheduleDisplayPassNow = NO;
+  static std::deque<ASDisplayNode *> __renderQueue = std::deque<ASDisplayNode *>();
+  static CFRunLoopObserverRef __mainRunLoopObserver = nil;
+  static NSUInteger __renderBatchSize = 1;
   {
     ASDN::MutexLocker l(__displaySchedulerLock);
-    
-    if (!__nodesToDisplay) {
-      __nodesToDisplay = [NSMutableArray array];
+
+    // Check if the node exists.
+    BOOL foundNode = NO;
+    for (ASDisplayNode *currentNode : __renderQueue) {
+      if (currentNode == node) {
+        foundNode = YES;
+        break;
+      }
     }
-    
-    if ([__nodesToDisplay indexOfObjectIdenticalTo:node] == NSNotFound) {
-      [__nodesToDisplay addObject:node];
+
+    if (!foundNode) {
+      __renderQueue.push_back(node);
     }
-    
-    if (!__displayScheduled) {
-      scheduleDisplayPassNow = YES;
-      __displayScheduled = YES;
+
+    if (!__mainRunLoopObserver) {
+
+      void (^handlerBlock) (CFRunLoopObserverRef observer, CFRunLoopActivity activity) = ^(CFRunLoopObserverRef observer, CFRunLoopActivity activity) {
+        std::deque<ASDisplayNode *> displayingNodes = std::deque<ASDisplayNode *>();
+        BOOL isQueueDrained = NO;
+        // Create a lock scope.
+        {
+          ASDN::MutexLocker l(__displaySchedulerLock);
+
+          // Early-exit if we don't have any nodes to render.
+          if (__renderQueue.empty()) {
+            return;
+          }
+
+          // Snatch the next batch of nodes.
+          NSUInteger totalNodeCount = __renderQueue.size();
+          for (int i = 0; i < MIN(__renderBatchSize, totalNodeCount); i++) {
+            ASDisplayNode *node = __renderQueue[0];
+            __renderQueue.pop_front();
+            displayingNodes.push_back(node);
+          }
+
+          if (__renderQueue.empty()) {
+            isQueueDrained = YES;
+          }
+        }
+
+        CFAbsoluteTime timestamp = isQueueDrained ? CFAbsoluteTimeGetCurrent() : 0;
```

Comment:
Should we get the timestamp while holding `__displaySchedulerLock`? It might be possible that one (or more) nodes are added right after we release the lock, but before we capture the timestamp, no?